File: algorithms/cpp/oddEvenJump/oddEvenJump.cpp
// Source: https://leetcode.com/problems/odd-even-jump/
// Author: Miao Zhang
// Date:   2021-03-30

class Solution {
public:
    int oddEvenJumps(vector<int>& arr) {
        int n = arr.size();
        map<int, int> m;
        vector<vector<int>> dp(n, vector<int>(2));
        dp[n - 1][0] = dp[n - 1][1] = 1;
        m[arr[n - 1]] = n - 1;
        int res = 1;
        for (int i = n - 2; i >= 0; i--) {
            auto o = m.lower_bound(arr[i]);
            if (o != m.end()) dp[i][0] = dp[o->second][1];
            auto e = m.upper_bound(arr[i]);
            if (e != m.begin()) dp[i][1] = dp[prev(e)->second][0];
            if (dp[i][0]) res++;
            m[arr[i]] = i;
        }
        return res;
    }
};

File: algorithms/python/oddEvenJump/oddEvenJump.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/odd-even-jump/
# Author: Miao Zhang
# Date:   2021-03-30

class Solution:
    def oddEvenJumps(self, arr: List[int]) -> int:
        n = len(arr)
        
        def make(arr1):
            res = [None] * n
            stack = []
            for num in arr1:
                while stack and num > stack[-1]:
                    res[stack.pop()] = num
                stack.append(num)
            return res
        
        arr1 = sorted(range(n), key=lambda i: arr[i])
        oddnext = make(arr1)
        arr1.sort(key=lambda i: -arr[i])
        evennext = make(arr1)
        
        odd = [False] * n
        even = [False] * n
        odd[n - 1] = even[n - 1] = True
        for i in range(n - 2, -1, -1):
            if oddnext[i] is not None:
                odd[i] = even[oddnext[i]]
            if evennext[i] is not None:
                even[i] = odd[evennext[i]]
        return sum(odd)
